idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/ui/popups/PopupSyntax.kt

```diff
@@ -0,0 +1,150 @@
+package arrow.meta.ide.dsl.ui.popups
+
+import arrow.meta.ide.IdeMetaPlugin
+import arrow.meta.ide.dsl.editor.action.AnActionSyntax
+import arrow.meta.ide.dsl.editor.inspection.InspectionSyntax
+import arrow.meta.internal.Noop
+import com.intellij.openapi.ui.MessageType
+import com.intellij.openapi.ui.popup.BalloonBuilder
+import com.intellij.openapi.ui.popup.ComponentPopupBuilder
+import com.intellij.openapi.ui.popup.JBPopup
+import com.intellij.openapi.ui.popup.JBPopupFactory
+import com.intellij.openapi.ui.popup.ListPopup
+import com.intellij.openapi.ui.popup.ListSeparator
+import com.intellij.openapi.ui.popup.PopupStep
+import com.intellij.openapi.ui.popup.PopupStep.FINAL_CHOICE
+import com.intellij.openapi.ui.popup.TreePopup
+import com.intellij.openapi.ui.popup.TreePopupStep
+import com.intellij.openapi.ui.popup.util.BaseListPopupStep
+import java.awt.Color
+import javax.swing.Icon
+import javax.swing.JComponent
+import javax.swing.event.HyperlinkEvent
+
+/**
+ * Popup's can be created at any time.
+ * Either in [AnActionSyntax], [InspectionSyntax] or [IntentionExtensionProviderSyntax] and many other places.
+ * Once the PopUp is created the user needs to call a `show()` method like [com.intellij.openapi.ui.popup.JBPopup.showUnderneathOf], this is an effectfull operation.
+ */
+interface PopupSyntax {
+  /**
+   * ```kotlin:ank:playground
+   * import arrow.meta.Plugin
+   * import arrow.meta.ide.IdeMetaPlugin
+   * import arrow.meta.ide.dsl.ui.popups.IdeListPopupItem
+   * import arrow.meta.ide.resources.ArrowIcons
+   * import arrow.meta.invoke
+   * import com.intellij.openapi.actionSystem.PlatformDataKeys
+   *
+   * val IdeMetaPlugin.showListPlugin: Plugin
+   *  get() = "Action to create ListPopUp" {
+   *    meta(
+   *      addAnAction(
+   *        "Unique",
+   *        anAction(
+   *          title = "Show ListPopUp",
+   *          description = "Shows one item of a ListPopUp",
+   *          icon = ArrowIcons.ICON4,
+   *          actionPerformed = { e ->
+   *            PlatformDataKeys.CONTEXT_COMPONENT.getData(e.dataContext)?.run {
+   *              listPopUp(
+   *                title = "TopLevelTitle",
+   *                sameIcon = ArrowIcons.ICON3, // every PopUp item will have this Icon
+   *                popUps = listOf(
+   *                  IdeListPopupItem(
+   *                    this,
+   *                    text = { "Teach your users about this Component: $this" }
+   *                  )
+   *                )
+   *              ).show(this)
+   *            }
+   *          }
+   *        )
+   *      )
+   *    )
+   *  }
+   * ```
```

Comment:
wow dang I can't believe how easy it is to write these! I love that it's the same user experience across the board. Great work Imran!